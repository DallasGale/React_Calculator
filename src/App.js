import React, { Component } from 'react';
import KeypadButton from './components/keypad/button';
import Display from './components/display';
import './App.css';

const numbers = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'];
const operators = [
  {
    'label': '+',
    'math': '+',
  },
  {
    'label': '-',
    'math': '-',
  },
  { 
    'label': 'รท',
    'math': '/'
  },
  {
    'label': 'ร', 
    'math': '*'
  }
];

class App extends Component {
  constructor() {
    super();
    this.state = {
      // state of numbers & characters as they are being typed
      input: [],
      // When an operator (+, -, /, *) has been 
      // entered disable the onClick() until a number has been entered

      operatorJustUsed: false, 
      totalled: false
    }
  }

  /**
   * 
   * @param { string } numeral 
   * '0-9 + .' numeral is added to the 
   * state.input array
   * 
   * @param { string } outputString 
   * convert enteredNumber to string and remove ','
   * 
   * */

  handleNumberInput = numeral => {
    this.setState({
      // add numeral to the input array
      input: [...this.state.input, numeral],
      operatorJustUsed: false
    })

    // When there is a total figure and you want to start 
    // adding new numbers that will overide the total
    if (this.state.totalled) {
      this.setState({
        input: [numeral],
        totalled: false
      })
    }
  }

  /**
   * 
   * @param { string } operator 
   * '+,-,/,*' operator is added to the state.input array
   * 
   * */

  handleOperatorInput = operator => {
    this.setState({
      input: this.state.input + operator,
      operatorJustUsed: true,
      totalled: false
    });
  }

  getTotal = () => {

    // Convert input array to a string, then remove the ',' 
    // to display as a complete numeral.
    // We don't convert the array to a number because there 
    // are mixed types (ie: 2,3 with /, *)

    let filter = this.state.input.toString().replace(/,/g, '');

    // Because the above filter will render a string of numbers 
    // plus operators (eg: '54+2/7'), we use the Function method
    // to return the string as if it was a javascript value (eg: 54+2/7).
    // See: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function

    let sum = Function(`return ${filter}`);
    // function sum(string) {
    //   const strip = string.replace(/'/g, '');
    //   console.log('strip: ', strip);
    //   return strip;
    // };

    // console.log('returned sum: ', sum(filter));

    this.setState({
      input: [sum().toString()],
      totalled: true,
      operatorJustUsed: false
    })
    console.log(this.state.input);
  }
  

  clear = () => {
    this.setState({
      input: [],
    })
  }

  handleNullClick = () => { return };

  componentDidUpdate() {
    console.log('this.state.input: ', this.state.input);
  }

  render() {
    const displayedInput = this.state.input.toString().replace(/,/g, '');
    return (
      <div className="App">
        <header className="App-header">
          <div className='app__wrapper'>
            {/* Calculator */}
            <Display value={displayedInput} />
            <div className='keypad'>
              <div className='keypad__numbers'>
                {
                  numbers.map((num) => (
                    <KeypadButton
                      buttonType='numbers'
                      id={num} key={num}
                      label={num}
                      onClick={() => this.handleNumberInput(num)} />
                  ))
                }
                <KeypadButton
                  buttonType='decimal'
                  id='.' key='.'
                  label='.'
                  onClick={() => this.handleNumberInput('.')} />

                <KeypadButton
                  buttonType='sum'
                  id='=' key='='
                  label='='
                  onClick={() => this.getTotal()} />
              </div>
              <div className='keypad__operators'>
                {
                  operators.map((op) => (
                    <KeypadButton
                      buttonType='operators'
                      id={op.label} key={op.label}
                      label={op.label}
                      onClick={() => this.state.operatorJustUsed ? this.handleNullClick() : this.handleOperatorInput(op.math)} />
                  ))
                }
                <KeypadButton
                  buttonType='operators'
                  id='C' key='C'
                  label='C'
                  onClick={() => this.clear()} />
              </div>
            </div>
          </div>
        </header>
      </div>
    );
  }
}


export default App;
