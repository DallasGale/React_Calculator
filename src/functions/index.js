//  Initial state of calculator
let initialState = {
    number: 0
}

export const keypadNumbers = [ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 ];

export const keypadControls = {
    'clear': 'C',
    'add': '+',
    'subtract': '-',
};

export const clear = () => {
    return initialState.number = 0;
};

export const add = (entered) => {
    return initialState.number += entered;
};

export const addOperator = () => {
    return '+';
};

export const divide = (entered) => {
    return initialState.number /= entered;
};

// export const handleKeypadNumerical = (initState, input, output) => {
//     output = parseInt(initState, 10) + parseInt(input, 10);
//     return output;
// };


export const handleKeypadNumerical = (input, output) => {
    return output + input;
};

export const handleMethodButtonClick = () => {
    return null;
}

export const multiply = (entered) => {
    return initialState.number *= entered;
};

export const setNumber = (entered) => {
    return initialState.number = entered;
};

export const subtract = (entered) => {
    return initialState.number -= entered;
};