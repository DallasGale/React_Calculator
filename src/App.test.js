/* eslint-disable no-unused-expressions */
import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, render } from 'enzyme';
import { add, addOperator, clear, divide, handleKeypadNumerical, multiply, setNumber, subtract } from './functions';
import KeypadButton from './components/keypad/button/';
import Display from './components/display';
import App from './App';

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
});


//////////////////////////////
// What needs to be tested: //
//////////////////////////////

// # + multiple times while adding to the total (done)
// # * multiple times while adding to the total (done)
// # - multiple times while adding to the total (done)
// # / multiple times while adding to the total (done)
// # Combine + and - (done)
// # Combing + and * (done)
// # Combine + and / (done)
// # Combine - and * (done)
// # Combine - and / (done)
// # Combine / and * (done)
// # Combine + and - and / and * (done)
// # Keypad onClick (done)
// # Keypad button click event
// # Reset/Clear  (done)
// # Round up to 2 decimal places
// # Render keypad control button labels (+, /, -, *, =, C)

//////////////////////////////

describe('Calculator methods', () => {
    test('clear calculator back to 0', () => {
        expect(clear()).toEqual(0);
        expect(clear()).not.toEqual(22)
    }),
        test('set initial number to anything (15)', () => {
            expect(setNumber(15)).toEqual(15);
        })

    test('Addition: 0 + 22', () => {
        expect(clear()).toEqual(0);
        expect(add(22)).toEqual(22);
    }),

        test('Addition: + multiple times while adding to the total', () => {
            expect(clear()).toEqual(0);
            expect(add(22)).toEqual(22);
            expect(add(3)).toEqual(25);
            expect(add(567)).toEqual(592);
        }),

        test('Division: Divide 10 / 2', () => {
            expect(setNumber(10)).toEqual(10);
            expect(divide(2)).toEqual(5);
        }),

        test('Division: / multiple times while adding to the total', () => {
            expect(setNumber(20)).toEqual(20);
            expect(divide(2)).toEqual(10);
            expect(divide(5)).toEqual(2);
        }),

        test('Multiplication: 10 * 4', () => {
            expect(setNumber(10)).toEqual(10);
            expect(multiply(4)).toEqual(40);
        }),

        test('Multiplication: * multiple times while adding to the total', () => {
            expect(setNumber(10)).toEqual(10);
            expect(multiply(4)).toEqual(40);
            expect(multiply(3)).toEqual(120);
        }),

        test('Subtraction: 54 - 110', () => {
            expect(setNumber(110)).toEqual(110);
            expect(subtract(54)).toEqual(56);
        }),

        test('Subtraction: - multiple times while adding to the total', () => {
            expect(setNumber(110)).toEqual(110);
            expect(subtract(54)).toEqual(56);
            expect(subtract(22)).toEqual(34);
        })
})

describe('Method combinations', () => {
    test('Combine addition & subtraction', () => {
        expect(setNumber(20)).toEqual(20);
        expect(add(45)).toEqual(65);
        expect(subtract(13)).toEqual(52);
    }),

        test('Combine addition & multiplication', () => {
            expect(setNumber(20)).toEqual(20);
            expect(add(45)).toEqual(65);
            expect(multiply(2)).toEqual(130);
        }),

        test('Combine addition & division', () => {
            expect(setNumber(20)).toEqual(20);
            expect(add(45)).toEqual(65);
            expect(divide(2)).toEqual(32.5);
        }),

        test('Combine subtraction & multiplication', () => {
            expect(setNumber(20)).toEqual(20);
            expect(subtract(45)).toEqual(-25);
            expect(multiply(2)).toEqual(-50);
        }),

        test('Combine subtraction & divide', () => {
            expect(setNumber(20)).toEqual(20);
            expect(subtract(45)).toEqual(-25);
            expect(divide(2)).toEqual(-12.5);
        }),

        test('Combine division & multiplication', () => {
            expect(setNumber(20)).toEqual(20);
            expect(divide(10)).toEqual(2);
            expect(multiply(10)).toEqual(20);
        }),

        test('Combine addition, subtraction, division & multiplication', () => {
            expect(clear()).toEqual(0);
            expect(add(1000)).toEqual(1000);
            expect(subtract(900)).toEqual(100);
            expect(divide(2)).toEqual(50);
            expect(multiply(8)).toEqual(400);
        })
})


describe('Keypad button', () => {

    it('should be defined', () => {
        expect(KeypadButton).toBeDefined();
    });

    it('should render correctly', () => {
        const tree = render(
            <KeypadButton
                id={0}
                key={0}
                label='C'
                name='clear'
                onClick={handleKeypadNumerical()} />
        );
        expect(tree).toMatchSnapshot();
    });
})

describe('Calculator display', () => {

    it('should render correctly', () => {
        let state = '0';
        const tree = render(
            <Display value={parseInt(state, 10)} />
        );
        expect(tree).toMatchSnapshot();
    });

    it('should display 0 at start', () => {
        let state = '0';
        const component = shallow(<Display value={parseInt(state, 10)} />)
        expect(component.text()).toEqual(state);
    });


    ///////////////////
    // initial state //
    ///////////////////
    const app = shallow(<App />);
    app.setState({
        displayNumber: '',
        displayNumberCopy: '',
        calculation: ''
    });
    let concatonatedState = app.state().displayNumber;
    let calculation = 0; // this is the variable that will return total (eg. 32+54+3-6/32);
    
    const DisplayComponent = shallow(
        <Display value={
            app.state.displayNumber
        } />
    )
    expect(DisplayComponent.text()).toEqual('');


    it('should concat a string of numbers', () => {
        expect(handleKeypadNumerical('8', concatonatedState)).toEqual('8');
        concatonatedState += '8';
        app.setState({ displayNumber: concatonatedState });
        expect(app.state().displayNumber).toEqual('8');

        expect(handleKeypadNumerical('4', concatonatedState)).toEqual('84');
        concatonatedState += '4'
        app.setState({ displayNumber: concatonatedState });
        expect(app.state().displayNumber).toEqual('84');

        expect(handleKeypadNumerical('9', concatonatedState)).toEqual('849');
        concatonatedState += '9'
        app.setState({ displayNumber: concatonatedState });
        expect(app.state().displayNumber).toEqual('849');
    });

    it('save concatonated state to a new variable', () => {
        calculation = + app.state().displayNumber;
        expect(calculation).toEqual(849);
    });

    it('assign an operator to the concatonated string', () => {
        expect(addOperator()).toEqual('+'); 
        app.setState({ displayNumber: app.state().displayNumber + '+' });
        expect(app.state().displayNumber).toEqual('849+');
    });

    it('should not be able to add concecutive operators', () => {
        // convert to string to array
        let arrayed = app.state().displayNumber.split('');
        expect(arrayed).toEqual(['8', '4', '9', '+']);
        let numberAfterOperator = false

        if ( arrayed.lastOfIndex('+') || arrayed.lastOfIndex('-') || arrayed.lastOfIndex('*') || arrayed.lastOfIndex('/') || arrayed.lastOfIndex('=')) {
            numberAfterOperator = true;
        }

        

        app.setState({ displayNumber: app.state().displayNumber + '*' });
        expect(app.state().displayNumber).toEqual('849+*');


    });

    it('clear concatonated state', () => {
        app.setState({ displayNumber: '' });
        expect(app.state().displayNumber).toEqual('');
        concatonatedState = '';
    });

    it('\'add\' new state to saved state', () => {

    });

    it('return total', () => {

    });


    it('clear original state; concat new state; \'add\' new state to saved state; return total', () => {

        ///////////////////
        // initial state //
        ///////////////////
        // const app = shallow(<App />);
        // app.setState({
        //     displayNumber: '',
        //     displayNumberCopy: '',
        //     calculation: ''
        // });

        ////////////////////////
        // concatonated state //
        ////////////////////////
        // let concatonatedState = app.state().displayNumber;
        // let calculation = 0; // this is the variable that will return total (eg. 32+54+3-6/32);

        /////////////////////////////////////
        // view component that shows state //
        /////////////////////////////////////
        // const DisplayComponent = shallow(
        //     <Display value={
        //         app.state.displayNumber
        //     } />
        // )
        // expect(DisplayComponent.text()).toEqual('');


        /////////////////////
        // UPDATE STATE... //
        /////////////////////
        // expect(handleKeypadNumerical('8', concatonatedState)).toEqual('8');
        // concatonatedState += '8';
        // app.setState({ displayNumber: concatonatedState });
        // expect(app.state().displayNumber).toEqual('8');

        // expect(handleKeypadNumerical('4', concatonatedState)).toEqual('84');
        // concatonatedState += '4'
        // app.setState({ displayNumber: concatonatedState });
        // expect(app.state().displayNumber).toEqual('84');

        // expect(handleKeypadNumerical('9', concatonatedState)).toEqual('849');
        // concatonatedState += '9'
        // app.setState({ displayNumber: concatonatedState });
        // expect(app.state().displayNumber).toEqual('849');


        /////////////////////////////////////////
        // SAVE STATE INTO 'CALCULATION' variable //
        /////////////////////////////////////////
        // app.setState({calculation: app.state().displayNumber});
        // calculation = + app.state().displayNumber;
        // expect(calculation).toEqual(849);


        ////////////////////////////////////////////////////
        // ASSIGN THE '+' OPERATOR TO 'CALCULATION' STATE //
        ////////////////////////////////////////////////////

        // expect(addOperator()).toEqual('+'); // this returns '+' to state.calculation
        // app.setState({ calculation: app.state().calculation + '+' });

        // expect(app.state().calculation).toEqual('+');
        // app.setState({calculation: app.state().calculation + '4'});
        // expect(app.state().calculation).toEqual('849+4');
        // let evalNumber = eval(app.state().calculation);
        // expect(evalNumber).toEqual(853);

        ///////////////////////////////////////////////
        // CLEAR CONCATONATED STATE //
        ///////////////////////////////////////////////
        // app.setState({ displayNumber: '' });
        // expect(app.state().displayNumber).toEqual('');
        // concatonatedState = '';


        //////////////////////
        // 'ADD' NEW NUMBER //
        //////////////////////
        // expect(handleKeypadNumerical('3', concatonatedState)).toEqual('3');
        // concatonatedState += '3';
        // app.setState({ displayNumber: concatonatedState });
        // expect(app.state().displayNumber).toEqual('3');

        // expect(handleKeypadNumerical('6', concatonatedState)).toEqual('36');
        // concatonatedState += '6'
        // app.setState({ displayNumber: concatonatedState });
        // expect(app.state().displayNumber).toEqual('36');


        /////////////////////////////////////////
        // SAVE STATE INTO 'CALCULATION' STATE //
        /////////////////////////////////////////
        // app.setState({calculation: app.state().calculation + app.state().displayNumber});
        // expect(app.state().calculation).toEqual('849+36');

        //////////////////////////////////
        // 'ADD' NEW STATE TO SAVED STATE //
        //////////////////////////////////
        // expect(handleKeypadNumerical('5', concatonatedState)).toEqual('5');
        // concatonatedState += '5';
        // app.setState({displayNumber: concatonatedState});
        // expect(app.state().displayNumber).toEqual('5');

        //////////////////////////////////
        // CONVERT 'STRING' TO 'NUMBER' //
        //////////////////////////////////
        // let rightNumber = parseInt(app.state().displayNumber, 10);
        // expect(rightNumber).toEqual(5)


        //////////////////
        // RETURN TOTAL //
        //////////////////
        // app.setState({calculation: leftNumber + rightNumber});
        // expect(app.state().calculation).toEqual(854);
    })
})