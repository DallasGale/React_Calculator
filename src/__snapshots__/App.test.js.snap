// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator display should render correctly 1`] = `
<div
  class="display__wrapper"
>
  <span
    class="display__text"
  >
    <span>
      0
    </span>
  </span>
</div>
`;

exports[`Keypad button should render correctly 1`] = `
<div
  class="keypad__button-wrapper  keypad__button-wrapper--undefined"
>
  <input
    class="keypad__button keypad__button--undefined"
    id="0"
    name="clear"
    type="button"
    value="C"
  />
</div>
`;
